batch/batch/driver/create_instance.py

```diff
@@ -178,6 +171,27 @@ def create_instance_config(
 DOCKER_ROOT_IMAGE=$(curl -s -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/attributes/docker_root_image")
 DOCKER_PREFIX=$(curl -s -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/attributes/docker_prefix")
 
+INTERNAL_GATEWAY_IP=$(curl -s -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/attributes/internal_ip")
+
+# private job network = 10.0.0.0/16
+# public job network = 10.1.0.0/16
+# [all networks] Rewrite traffic coming from containers to masquerade as the host
+iptables --table nat --append POSTROUTING --source 10.0.0.0/15 --jump MASQUERADE
+
+# [public]
+# Block public traffic to the metadata server
+iptables --append FORWARD --source 10.1.0.0/16 --destination 169.254.169.254 --jump DROP
+# But allow the internal gateway
+iptables --append FORWARD --destination $INTERNAL_GATEWAY_IP --jump ACCEPT
+# And this worker
+iptables --append FORWARD --destination $IP_ADDRESS --jump ACCEPT
+# For public jobs, forbid outgoing requests to cluster-internal IP addresses. For not-public jobs, allow
+# all outgoing packets
+ENS_DEVICE=$(ip link list | grep ens | awk -F": " '{{ print $2 }}')
+iptables --append FORWARD --source 10.1.0.0/16 --out-interface $ENS_DEVICE ! --destination 10.128.0.0/16 --jump ACCEPT
+iptables --append FORWARD ! --source 10.1.0.0/16 --out-interface $ENS_DEVICE --jump ACCEPT
```

Comment:
Let me make sure I understand. This says forward any request not from the public network to any device in our network. I'm not sure what ENS_DEVICE means, but I assume it translate to our cluster.

Likewise, `! --denstination 10.128.0.0/16` is preventing any packet not destined to our internal IP. 

Why are both $ENS_DEVICE and `10.128.0.0/16` needed?

Reply:
`ENS_DEVICE` is the logical equivalent of the ethernet port that connects the worker machine to the network. The reason I'm specifying this device is to distinguish between traffic that is leaving the worker and traffic inside the worker, e.g. if a job was trying to contact another colocated job.

The first iptables rule  *permits* public traffic leaving the worker (going through `ENS_DEVICE`) so long as it is *not* destined for another node on our network (`10.128.0.0./16`). Without the `destination`, we would be permitting any public outgoing traffic, including to other internal IPs. Without the `out-interface`, we would be permitting any traffic not going to other internal IPs, which means that a job could talk to another job on the same worker since jobs are given `10.0.x.x` or `10.1.x.x` IPs.

Reply:
Correct about the second iptables rule. *Not* from the public network essentially just means private network + host.